File: app-py/src/hanik_mcp/identity/caller_identity.py
"""Identity context — ``CallerIdentity`` and its resolver.

The resolver is an Anti-Corruption Layer: it translates the two foreign
authentication mechanisms that can front this server into a single domain
identity, so the rest of the code never touches raw Easy Auth headers or JWTs.

 1. Azure Container Apps "Easy Auth" injected headers (x-ms-client-principal*).
 2. A forwarded Entra ID bearer token in the Authorization header.
"""

from __future__ import annotations

import base64
import json
from typing import Any, Mapping

from .access_token import AccessToken


def resolve_caller_identity(headers: Mapping[str, str]) -> dict[str, Any]:
    def header(name: str) -> str | None:
        value = headers.get(name)
        return value if isinstance(value, str) else None

    # 1) Easy Auth simple headers.
    principal_name = header("x-ms-client-principal-name")
    principal_id = header("x-ms-client-principal-id")
    principal_idp = header("x-ms-client-principal-idp")

    # 1b) Easy Auth full base64-encoded principal (claims array).
    easy_auth_claims: dict | None = None
    encoded_principal = header("x-ms-client-principal")
    if encoded_principal:
        try:
            easy_auth_claims = json.loads(
                base64.b64decode(encoded_principal).decode("utf-8")
            )
        except (ValueError, json.JSONDecodeError):
            easy_auth_claims = None

    # 2) Bearer token claims.
    token = AccessToken.from_authorization_header(header("authorization"))
    claims: dict = token.claims if token and token.claims else {}

    display_name = (
        principal_name
        or claims.get("name")
        or claims.get("preferred_username")
        or claims.get("upn")
    )
    user_id = principal_id or claims.get("oid") or claims.get("sub")

    authenticated = bool(display_name or user_id or easy_auth_claims)

    return {
        "authenticated": authenticated,
        "displayName": display_name or None,
        "userPrincipalName": (
            claims.get("preferred_username") or claims.get("upn") or principal_name
        ),
        "email": claims.get("email") or None,
        "objectId": user_id or None,
        "tenantId": claims.get("tid") or None,
        "identityProvider": principal_idp,
        "scopes": claims.get("scp") or claims.get("roles") or None,
    }

File: app/src/identity/CallerIdentity.ts
import type { Request } from "express";
import { AccessToken } from "./AccessToken.js";

/**
 * Identity context — `CallerIdentity` Value Object and its resolver.
 *
 * The resolver is an Anti-Corruption Layer: it translates the two foreign
 * authentication mechanisms that can front this server into a single domain
 * identity, so the rest of the code never touches raw Easy Auth headers or JWTs.
 *  1. Azure Container Apps "Easy Auth" injected headers (x-ms-client-principal*).
 *  2. A forwarded Entra ID bearer token in the Authorization header.
 */
export interface CallerIdentity {
  authenticated: boolean;
  displayName: string | null;
  userPrincipalName: string | null;
  email: string | null;
  objectId: string | null;
  tenantId: string | null;
  identityProvider: string | null;
  scopes: string | string[] | null;
}

export function resolveCallerIdentity(req: Request): CallerIdentity {
  const headers = req.headers;

  // 1) Easy Auth simple headers.
  const principalName = headers["x-ms-client-principal-name"];
  const principalId = headers["x-ms-client-principal-id"];
  const principalIdp = headers["x-ms-client-principal-idp"];

  // 1b) Easy Auth full base64-encoded principal (claims array).
  let easyAuthClaims: Record<string, unknown> | null = null;
  const encodedPrincipal = headers["x-ms-client-principal"];
  if (typeof encodedPrincipal === "string") {
    try {
      easyAuthClaims = JSON.parse(
        Buffer.from(encodedPrincipal, "base64").toString("utf8")
      ) as Record<string, unknown>;
    } catch {
      easyAuthClaims = null;
    }
  }

  // 2) Bearer token claims.
  const token = AccessToken.fromAuthorizationHeader(headers["authorization"]);
  const claims = token?.claims ?? {};

  const displayName =
    (typeof principalName === "string" ? principalName : undefined) ??
    (claims.name as string | undefined) ??
    (claims.preferred_username as string | undefined) ??
    (claims.upn as string | undefined);
  const userId =
    (typeof principalId === "string" ? principalId : undefined) ??
    (claims.oid as string | undefined) ??
    (claims.sub as string | undefined);

  const authenticated = Boolean(displayName || userId || easyAuthClaims);

  return {
    authenticated,
    displayName: displayName ?? null,
    userPrincipalName:
      (claims.preferred_username as string | undefined) ??
      (claims.upn as string | undefined) ??
      (typeof principalName === "string" ? principalName : null),
    email: (claims.email as string | undefined) ?? null,
    objectId: userId ?? null,
    tenantId: (claims.tid as string | undefined) ?? null,
    identityProvider: typeof principalIdp === "string" ? principalIdp : null,
    scopes:
      (claims.scp as string | undefined) ??
      (claims.roles as string[] | undefined) ??
      null,
  };
}
